File: problem_0015.py
#!/usr/bin/python
"""Starting in the top left corner of a 2x2 grid, and only being able to move to the right and down, there are exactly 6 routes to the bottom right corner.

How many such routes are there through a 20x20 grid?"""

import time

def main():
    print new_method(20,20)
#   node_map = create_map(10,10)
#   print unique_paths(len(node_map)-1, node_map)

def new_method(width,height):
    lines = [range(2,width+2)]
    for i in range(3, height+2):
        line = []
        for j,num in enumerate(lines[-1]): 
            if j == 0: value = i 
            else     : value = line[-1] + num
            line.append(value)
        lines.append(line)
    return lines[-1][-1]

def create_map(width, height):
    previous_line = [None] * width
    previous_node = None
    node_map = []
    for h in range(height+1):
        for w in range(width+1):
            if h == 0: down  = None
            else     : down  = len(node_map)-width-1
            if w == 0: right = None
            else     : right = len(node_map)-1
            node_map.append((down,right))
    return node_map

def unique_paths(root, node_map):
    queue = [root]
    path_count = 0
    while queue:
        node = node_map[queue.pop()]
        has_steps = False
        for step in node:
            if step:
                has_steps = True
                queue.append(step)
        if not has_steps:
            path_count += 1
    return path_count

if __name__ == "__main__":
    start = time.time()
    main()
    print "That took %0.5f seconds" % (time.time() - start)

File: problem_0015.hs
{-# LANGUAGE DeriveDataTypeable, RecordWildCards #-}

import Text.Printf (printf)
import Test.HUnit ((@=?), runTestTT, Test(..))
import System.Console.CmdArgs
import Data.Time (getCurrentTime, diffUTCTime)
import System.IO (IOMode( ReadMode), hGetContents, openFile)
import Data.Char (digitToInt)

-- Starting in the top left corner of a 2×2 grid, and only being able to move to the right and down, there are exactly 6 routes to the bottom right corner.

-- How many such routes are there through a 20×20 grid?

problem0015 :: (Int, Int) -> Integer
problem0015 (x, y) = paths !! x !! y

paths :: [[Integer]]
paths = firstRow : [ otherRows y | y <- [1..]]
    where
        firstRow = 0 : [1,1..]
        otherRows y = 1 : [ oneLeft x y + oneUp x y | x <- [1..]]
        oneUp x y = paths !! (x - 1) !! y
        oneLeft x y = paths !! x !! (y -1)

pathsTest :: [Test]
pathsTest = map TestCase [
    0 @=? paths !! 0 !! 0, 
    1 @=? paths !! 1 !! 0, 
    1 @=? paths !! 0 !! 1
    ]

problem0015Test :: [Test]
problem0015Test = map TestCase [
    0 @=? problem0015 ( 0,  0),
    1 @=? problem0015 (17,  0),
    1 @=? problem0015 ( 0, 16),
    6 @=? problem0015 ( 2,  2)
    ]

data EulerArgs = 
    AdHoc { gridSize :: (Int, Int) }
    | Euler 
    | UnitTest
        deriving (Show, Data, Typeable)

exec :: EulerArgs -> IO ()
exec AdHoc{..} = do
    let answer = problem0015 gridSize
        x = fst gridSize
        y = snd gridSize
    printf "In a %d x %d grid there are %d paths\n" x y answer
    printf "from the upper left to the lower right\n"
exec Euler = do
    let answer = problem0015 (20, 20)
    printf "Answer: %d\n" answer 
exec UnitTest = do 
    runTestTT $ TestList $ problem0015Test ++ pathsTest
    return ()

adHoc = AdHoc{ gridSize = (2,2) }
unittest = UnitTest{}
euler = Euler{}

main :: IO ()
main = do
    args <- cmdArgs $ modes [euler, unittest, adHoc]
    start <- getCurrentTime
    exec args
    stop <- getCurrentTime
    print $ diffUTCTime stop start
